metaswitch/common/generate_stats_csv.py

```diff
@@ -0,0 +1,266 @@
+
+
+def setup_logging():
+    """Set up basic logging."""
+    level = os.getenv('LOG_LEVEL', 'WARNING')
+    logging.basicConfig(level=getattr(logging, level))
+
+
+def parse_args():
+    """Get the command-line arguments."""
+    parser = argparse.ArgumentParser(description=__doc__)
+    parser.add_argument('mib_files', metavar='MIB', nargs='+',
+                        help='The absolute path(s) of the input MIB(s).')
+    parser.add_argument('csv_file', metavar='CSV',
+                        help='The absolute path of the input CSV.')
+    parser.add_argument('--output-file', default='./output.csv',
+                        help='Optional output file name (defaults to '
+                        'output.csv in the current directory).')
+    args = vars(parser.parse_args())
+    logger.debug("Command-line arguments: %s", args)
+    return args
+
+
+def parse_mib_files(mib_files):
+    """Parse the specified MIB files.
+
+    `mib_files` should be an iterator of absolute paths to ASN.1 MIB files on
+    disk.
+    Returns a dict keyed by (MIB table name, MIB field name) with values
+    (Source File, MIB table description, OID, MIB field description).
+    """
+    output = {}
+    for mib_file in mib_files:
+        output.update(parse_mib_file(mib_file))
+    logging.debug("Parsed MIBs: %s", output)
+    return output
+
+
+def parse_mib_file(path):
+    """Parse the specified MIB file.
+
+    `path` should be the absolute path to an ASN.1 MIB file on disk.
+    Returns a dict keyed by (MIB table name, MIB field name) with values
+    MibData entries.
+    """
+    mib_file = mib.MibFile(path)
+
+    def stat_test(stat):
+        """Filter for leaf nodes in OID tree."""
+        # TODO: pick the right nodes!
+        try:
+            stat.table()
+        except LookupError:
+            return False
+        return True
+
+    columns = ["SNMP NAME", "SOURCE FILE", "DESCRIPTION", "OID", ]
+    stats = mib_file.get_all_stats(columns).values()
+
+    leaf_stats = [stat for stat in stats if stat_test(stat)]
+
+    def key(stat):
+        return KeyData(mib_table_name=stat.table().get_info("SNMP NAME"),
+                       mib_field_name=stat.get_info("SNMP NAME"))
+
+    def value(stat):
+        return MibData(
+            source_file=stat.get_info("SOURCE FILE"),
+            mib_table_description=stat.table().get_info("DESCRIPTION"),
+            mib_field_description=stat.get_info("DESCRIPTION"),
+            oid=stat.get_info("OID"),
+        )
+
+    mib_dict = {key(stat): value(stat) for stat in leaf_stats}
+
+    logger.debug("Parsed MIB file, found stats %s", mib_dict)
+    return mib_dict
+
+
+def parse_csv_file(csv_file):
+    """Parse the specified CSV file.
+
+    `csv_file` should be an absolute path to a CSV file on disc.
+    Returns an OrderedDict keyed by (MIB table name, MIB field name) with
+    values CsvData entries.
+    """
+    with open(csv_file, 'rb') as csvfile:

```

Comment:
Would be good to handle a missing file a bit more nicely. Currently all you know is that there are a bunch of missing stats.